internal/mesh/internal/controllers/sidecarproxy/cache/cache.go

```diff
@@ -0,0 +1,144 @@
+package cache
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/consul/internal/mesh/internal/types/intermediate"
+	"github.com/hashicorp/consul/internal/resource"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+// Cache stores information needed for the sidecar-proxy controller to reconcile efficiently.
+// This currently means storing a list of all destinations for easy look up
+// as well as indices of source proxies where those destinations are referenced.
+//
+// It is the responsibility of the controller and its subcomponents (like mapper and data fetcher)
+// to keep this cache up-to-date as we're observing new data.
+type Cache struct {
+	lock sync.RWMutex
+
+	// store is a map from destination service reference and port as a string ID
+	// to the object representing destination reference.
+	store map[string]*intermediate.CombinedDestinationRef
+
+	// sourceProxiesIndex stores a map from a string representation of source proxy ID
+	// to the keys in the store map.
+	sourceProxiesIndex map[string]storeKeys
+}
+
+type storeKeys map[string]struct{}
+
+func New() *Cache {
+	return &Cache{
+		store:              make(map[string]*intermediate.CombinedDestinationRef),
+		sourceProxiesIndex: make(map[string]storeKeys),
+	}
+}
+
+func KeyFromID(id *pbresource.ID) string {
+	return fmt.Sprintf("%s/%s/%s",
+		resource.ToGVK(id.Type),
+		resource.TenancyToString(id.Tenancy),
+		id.Name)
+}
+
+func KeyFromRefAndPort(ref *pbresource.Reference, port string) string {
+	return fmt.Sprintf("%s:%s",
+		resource.ReferenceToString(ref),
+		port)
+}
+
+func (c *Cache) Write(d *intermediate.CombinedDestinationRef) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	key := KeyFromRefAndPort(d.ServiceRef, d.Port)
+
+	c.store[key] = d
+
+	// Update source proxies index.
+	for _, proxyID := range d.SourceProxies {
+		proxyIDKey := KeyFromID(proxyID)
+
+		_, ok := c.sourceProxiesIndex[proxyIDKey]
+		if !ok {
+			c.sourceProxiesIndex[proxyIDKey] = make(storeKeys)
+		}
+
+		c.sourceProxiesIndex[proxyIDKey][key] = struct{}{}
+	}
+}
+
+func (c *Cache) Delete(ref *pbresource.Reference, port string) {
```

Comment:
I found it valuable to do precondition-style assertions about incoming Types in deeply internal plumbing functions like these. So for example here you'd do:

```
if !resource.EqualType(types.EXPECTED, ref.Type) {
  panic("incorrect type")
}
```

That way you notice a hard programming error far earlier.